File: src/main/java/dev/kensa/output/template/Template.java
package dev.kensa.output.template;

import com.eclipsesource.json.Json;
import com.mitchellbosecke.pebble.PebbleEngine;
import com.mitchellbosecke.pebble.loader.ClasspathLoader;
import com.mitchellbosecke.pebble.template.PebbleTemplate;
import dev.kensa.KensaException;
import dev.kensa.context.TestContainer;
import dev.kensa.render.Renderers;

import java.io.IOException;
import java.net.URL;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.BiFunction;

import static dev.kensa.output.json.JsonTransforms.*;
import static java.nio.charset.StandardCharsets.UTF_8;
import static java.nio.file.StandardOpenOption.CREATE;

public class Template {

    public enum Mode {
        SingleFile,
        MultiFile,
        TestFile,
        Site
    }

    private static final PebbleEngine pebbleEngine = new PebbleEngine.Builder().autoEscaping(false).loader(new ClasspathLoader()).build();

    private final List<Index> indices = new ArrayList<>();
    private final PebbleTemplate template;
    private final List<JsonScript> scripts = new ArrayList<>();
    private final Path outputPath;

    private int indexCounter = 1;

    public Template(Path outputPath, Mode mode, URL issueTrackerUrl) {
        this.outputPath = outputPath;
        this.template = pebbleEngine.getTemplate("pebble-index.html");

        scripts.add(configurationJson(mode, issueTrackerUrl));
    }

    public void addIndex(TestContainer container, BiFunction<TestContainer, Integer, Index> factory) {
        indices.add(factory.apply(container, indexCounter++));
    }

    public void addJsonScript(TestContainer container, BiFunction<TestContainer, Integer, JsonScript> factory) {
        scripts.add(factory.apply(container, indexCounter));
    }

    public void write() {
        Map<String, Object> context = new HashMap<>();
        context.put("scripts", scripts);
        context.put("indices", indices);
        write(context);
        this.scripts.clear();
        indices.clear();
    }

    private void write(Map<String, Object> context) {
        try {
            template.evaluate(Files.newBufferedWriter(outputPath, UTF_8, CREATE), context);
        } catch (IOException e) {
            throw new KensaException("Unable to write template", e);
        }
    }

    private JsonScript configurationJson(Mode mode, URL issueTrackerUrl) {
        return new JsonScript(
                "config",
                toJsonString()
                        .apply(
                                Json.object()
                                    .add("mode", mode.name())
                                    .add("issueTrackerUrl", issueTrackerUrl == null ? "" : issueTrackerUrl.toString())
                        )
        );
    }

    public static BiFunction<TestContainer, Integer, Index> asIndex() {
        return (container, index) -> container.transform(toIndexJson("test-result-" + index)
                                                                 .andThen(toJsonString())
                                                                 .andThen(Index::new)
        );
    }

    public static BiFunction<TestContainer, Integer, JsonScript> asJsonScript(Renderers renderers) {
        return (container, index) -> container.transform(toJsonWith(renderers)
                                                                 .andThen(toJsonString())
                                                                 .andThen(js -> new JsonScript("test-result-" + index, js))
        );
    }
}

File: src/main/java/dev/kensa/output/template/Template.kt
package dev.kensa.output.template

import com.eclipsesource.json.Json
import dev.kensa.Configuration
import dev.kensa.KensaException
import dev.kensa.context.TestContainer
import dev.kensa.output.json.JsonTransforms.andThen
import dev.kensa.output.json.JsonTransforms.toIndexJson
import dev.kensa.output.json.JsonTransforms.toJsonString
import dev.kensa.output.json.JsonTransforms.toJsonWith
import dev.kensa.output.template.FileTemplate.Mode.IndexFile
import dev.kensa.output.template.FileTemplate.Mode.TestFile
import dev.kensa.sentence.Acronym
import io.pebbletemplates.pebble.PebbleEngine
import io.pebbletemplates.pebble.loader.ClasspathLoader
import io.pebbletemplates.pebble.template.PebbleTemplate
import java.io.IOException
import java.nio.charset.StandardCharsets.UTF_8
import java.nio.file.Files.newBufferedWriter
import java.nio.file.Path
import java.nio.file.StandardOpenOption.CREATE
import kotlin.io.path.createDirectories
import com.eclipsesource.json.Json.`object` as jsonObject

class JsonScript(@Suppress("unused", "for pebble template") val id: String, @Suppress("unused", "for pebble template") val content: String)
class Index(@Suppress("unused", "for pebble template") val content: String)

sealed class FileTemplate(mode: Mode, configuration: Configuration, private val templateOutputPath: Path) {

    enum class Mode { IndexFile, TestFile }

    private val template: PebbleTemplate = pebbleEngine.getTemplate("pebble-index.html")
    private val kensaScript: String = configuration.outputDir.resolve("kensa.js").toString()

    private val templateMap = mutableMapOf<String, MutableList<Any>>("scripts" to mutableListOf<Any>(configuration.asJson(mode)))

    protected fun add(key: String, value: Any) {
        templateMap.getOrPut(key) { mutableListOf<Any>() }.add(value)
    }

    fun write() {
        try {
            template.evaluate(
                newBufferedWriter(templateOutputPath, UTF_8, CREATE),
                mapOf("kensaScript" to kensaScript) + templateMap
            )
        } catch (e: IOException) {
            throw KensaException("Unable to write template", e)
        }
    }

    class IndexFileTemplate(configuration: Configuration) : FileTemplate(IndexFile, configuration, configuration.outputDir.resolve("index.html")) {

        private var indexCounter = 1

        fun addIndex(container: TestContainer) {
            add("indices", container.transform(toIndexJson("test-result-$indexCounter++").andThen(toJsonString()).andThen(::Index)))
        }
    }

    class TestFileTemplate(configuration: Configuration, container: TestContainer) : FileTemplate(TestFile, configuration, container.deriveTestPath(configuration.outputDir, configuration.flattenOutputPackages)) {

        init {
            add("scripts", container.transform(toJsonWith(configuration.renderers).andThen(toJsonString().andThen { content: String -> JsonScript("test-result-1", content) })))
        }

        companion object {

            private fun TestContainer.deriveTestPath(outputDir: Path, flattenOutputPackages: Boolean): Path =
                if (flattenOutputPackages) {
                    outputDir.resolve("${testClass.name}.html")
                } else {
                    createDirectoriesAndCalculatePath(outputDir, testClass.name)
                }

            private fun createDirectoriesAndCalculatePath(outputDir: Path, fullClassName: String): Path {
                val parts = fullClassName.split(".")

                return outputDir.resolve(parts.dropLast(1).joinToString(separator = "/")).run {
                    createDirectories()

                    resolve("${parts.last()}.html")
                }
            }
        }
    }

    companion object {
        private val pebbleEngine = PebbleEngine.Builder().autoEscaping(false).loader(ClasspathLoader()).build()
        private fun Configuration.asJson(mode: Mode): JsonScript =
            JsonScript(
                "config",
                toJsonString()(
                    jsonObject()
                        .add("mode", mode.name)
                        .add("issueTrackerUrl", issueTrackerUrl.toString())
                        .add("acronyms", acronymsAsJson(acronyms))
                        .add("flattenPackages", flattenOutputPackages)
                        .add("sectionOrder", Json.array().apply {
                            sectionOrder.forEach { add(it.name) }
                        })
                )
            )

        private fun acronymsAsJson(acronyms: Collection<Acronym>) = jsonObject().apply {
            acronyms.forEach {
                add(it.acronym, it.meaning)
            }
        }
    }
}